### PR TITLE
Nullable V2 with Generics

### PR DESCRIPTION
**Pull request by jonathansharman**
### Depends On

- To be tested via https://github.com/nicheinc/entity/pull/208 (private to the Niche GitHub organization)

### Documentation

- [Version 2 features and migration guide](https://github.com/nicheinc/nullable/blob/feature/generics/docs/v2-migration-guide.md)

### Description

This PR upgrades the library to version 2, with go 1.18 generics and refactored types and methods. Please see the [draft migration guide](https://github.com/nicheinc/nullable/blob/feature/generics/docs/v2-migration-guide.md) for a detailed overview of what has changed. (Feedback on that document is also of course very welcome.) Besides the conversion to generics, my overall goal was to make these types safer and more explicit and to lean harder into the nullable update abstraction (this library provides types that represent _updates to fields of go structs_). I'd be happy to discuss any of the design/naming decisions if you're interested.

I regretted calling this module/package an eight-character name, which would be especially cumbersome now that we often need type annotations. Therefore I've renamed the package to `nup` (for "nullable update"), which is short, memorable, pronounceable, and not taken by existing libraries. The problem is that the package and module names no longer match. These are the options I could think of for addressing this:

1. Move package `nup` into a subdirectory. Imports look fairly clean (`"github.com/nicheinc/nullable/nup/v2"`), and the VS Code Go extension can still auto-import the package when using its types. However, I think this is an unusual structure for the main (currently only) package of a module.
2. Leave package `nup` at the top of the repo. Imports will have to use `nup "github.com/nicheinc/nullable/v2"`, which is a little confusing/ugly/inconvenient. The VS Code Go extension will not automatically import the package when using its types.
3. Revert the package name back to `nullable`. Type names become annoyingly long again.
4. Make a new repo called `github.com/nicheinc/nup`. Mark the old repo as deprecated and point to the new one. In this case, the new library would stay at `v1`, I guess?
5. Rename the module and repo to `github.com/nicheinc/nup` (or still `.../v2`?). I assume this would break anything that currently depends on `nullable`, which is a nonstarter.

I went with option 1, but I'm open to other solutions if any reviewers disagree with that choice.

### Testing Considerations

This will be QA'd via this `entity` service branch (which is private to the Niche GitHub organization): https://github.com/nicheinc/entity/pull/208

### Versioning

Major.

### CODE REVIEW

File: nup/marshalJSON.go [hunk 1]
```diff
@@ -69,10 +77,10 @@ func MarshalJSON(v interface{}) ([]byte, error) {
 			return nil
 		}
 		switch field := fieldValue.Addr().Interface().(type) {
-		case Nullable:
-			// Only marshal nullable fields that are explicitly set.
-			if field.IsSet() {
-				if err := appendField(field.InterfaceValue()); err != nil {
+		case updateMarshaller:
```
Comment (jonathansharman) [inline nup/marshalJSON.go:80]: Users no longer need to implement new update types, so we don't need `Nullable` anymore. However, we still need an unexported interface for marshalling purposes, just to checking if an update field is a set/remove operation and getting its value as an `interface{}`.

I would like to just match `Update` and `SliceUpdate` values and skip this rather meaningless interface, but as far as I can tell there's currently no way to match against generic types in a type switch statement. You can match against an _instance_ of a generic type, e.g. `Update[int]`, but not against the generic type itself, e.g. `Update`/`Update[T any]`. Maybe a future go version will add support for that kind of thing.

File: nup/marshalJSON.go [hunk 2]
```diff
@@ -43,7 +51,7 @@ func MarshalJSON(v interface{}) ([]byte, error) {
 			continue
 		}
 		appendField := func(fieldInterface interface{}) error {
-			valueBuf, err := json.Marshal(fieldInterface)
+			valueBuf, err := MarshalJSON(fieldInterface)
```
Comment (jonathansharman) [inline nup/marshalJSON.go:54]: This change enables omitting nested no-op updates.

File: nup/update.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+// IsSetTo returns whether the update sets to the given value.
+func (u Update[T]) IsSetTo(value T) bool {
+	return u.op == OpSet && u.value == value
+}
+
+// IsSetSuchThat returns whether the update is a set operation to a value that
+// satisfies the given predicate.
+func (u Update[T]) IsSetSuchThat(predicate func(T) bool) bool {
+	return u.op == OpSet && predicate(u.value)
+}
```
Comment (jonathansharman) [inline nup/update.go:148]: I ended up not using this method anywhere in my `entity` branch, mainly because anonymous functions are pretty bulky in go. I think this still might sometimes be handy for one-off checks though, especially when an existing function can be passed in.

File: nup/update.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+func (u Update[T]) String() string {
+	switch u.op {
+	case OpNoop:
+		return "<no-op>"
+	case OpRemove:
+		return "<remove>"
+	}
+	switch value := interface{}(u.value).(type) {
+	case string:
+		return value
```
Comment (jonathansharman) [inline nup/update.go:161]: This case isn't necessary to get the desired behavior, but it can avoid a `Sprintf` call.

File: docs/v2-migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+`%v` since a `[]T` is not `string` or `fmt.Stringer`.)
+
+# Migration Guide
+
+The package has been renamed from `nullable` to `nup` (for "nullable update").
+It's 62.5% shorter and 87% more whimsical! You should replace v1 imports
+(`github.com/nicheinc/nullable`) with `github.com/nicheinc/nullable/v2/nup`.
+
+The old `Value` method on each `Nullable` type has been replaced with three
+`Update` methods: `Value`, `ValueOperation`, and `ValueOrNil`:
```
Comment (jonathansharman) [inline docs/v2-migration-guide.md:31]: My rationale behind these three methods is to make value accesses more explicit and less prone to panics. `Value` now returns two values, and since unused local variables are an error, the user has to explicitly ignore the value with `_` to "assert" that the value exists, rather than just dereferencing a pointer. `ValueOperation` is similar. I originally wasn't going to include `ValueOrNil` at all, but it was too useful in our existing patterns (especially for passing updates as postgres parameters).

**Review by dzbee — CHANGES_REQUESTED** [review]
It seems like we don't plan to update the repo name, is that correct? I can imagine possible reasons to not do it, but I'm curious what considerations you've given to it.

My biggest concern with this is that `SliceSet[T](nil)` is a valid function call that in many ways functions as `SliceRemove[T]()` but in some ways does not, and I think we should probably make it unambiguous that `SliceSet[T](nil)` should not be used for constructing remove operations to avoid potential gotchas (which already show up in round trip JSON flows).

File: nup/update_test.go [hunk 6]
```diff
@@ -0,0 +1,472 @@
+package nup
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/update_test.go:1]: I'm seeing 100% test coverage - are these methods not showing as covered for you? `Operation` is tested via [`TestUpdate_OperationAccessors`](https://github.com/nicheinc/nullable/blob/0b2d0709c8cbe5642ec4b99bece2210afca7ff6a/nup/update_test.go#L165) (I just grouped all the operation accessor methods into one test function). `interfaceValue` is tested via its use in `MarshalJSON`. In general, I prefer to unit test unexported helper methods via exported types/methods - in other words to test through the public interface.

File: nup/update.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	case fmt.Stringer:
+		return value.String()
+	default:
+		return fmt.Sprintf("%v", value)
+	}
+}
+
+// interfaceValue partially implements updateMarshaller.
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/update.go]: I like for struct method docs to mostly point to the interface docs (to reduce duplication), but I agree this is a little sparse. Clarified in 6e462424e82b88622449a920c661355f5d700e1a and 53e7017133f9613da588958249b97cfd3001e0d5.

File: nup/sliceUpdate_test.go [hunk 8]
```diff
@@ -0,0 +1,464 @@
+package nup
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/sliceUpdate_test.go:1]: See response: https://github.com/nicheinc/nullable/pull/9#discussion_r862154127

File: nup/sliceUpdate_test.go [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+		expected bool
+	}{
+		{
+			name:     "Noop",
+			u:        SliceNoop[int](),
+			expected: false,
+		},
+		{
+			name:     "Remove",
+			u:        SliceSet[int](nil),
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/sliceUpdate_test.go]: Fixed in 636cd1b245edb0e81fc2d6fe2a2eb4355c51cbfb

File: nup/sliceUpdate_test.go [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			if actual := testCase.u.IsSetSuchThat(isCouple); actual != testCase.expected {
+				t.Errorf("Expected: %v. Actual: %v", testCase.expected, actual)
+			}
+		})
+	}
+}
+
+func TestSliceUpdate_String(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/sliceUpdate_test.go:429]: Added in 636cd1b245edb0e81fc2d6fe2a2eb4355c51cbfb. In the case of `String` output, the ambiguity shifts from "can't tell if remove or set-`nil`" to "can't tell if set-`nil` or set-empty"., which I think is fine.

File: nup/marshalJSON_test.go [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+		Field Update[int]
+	}{
+		Field: Set(1),
+	})
+	roundtrip(t, "SliceUpdate/Noop", struct {
+		Field SliceUpdate[int]
+	}{
+		Field: SliceNoop[int](),
+	})
+	roundtrip(t, "SliceUpdate/Remove", struct {
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/marshalJSON_test.go:271]: Added a disclaimer/justification comment in 636cd1b245edb0e81fc2d6fe2a2eb4355c51cbfb

File: nup/sliceUpdate_test.go [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+		expected bool
+	}{
+		{
+			name:     "Noop",
+			u:        SliceNoop[int](),
+			expected: false,
+		},
+		{
+			name:     "Remove",
+			u:        SliceSet[int](nil),
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/sliceUpdate_test.go]: Very good point! I fixed those cases and added some extra tests and a comment concerning this ambiguity in 636cd1b245edb0e81fc2d6fe2a2eb4355c51cbfb. In short, I think we actually don't need to worry about it.

Removing a slice field is semantically equivalent to setting it to `nil`. This is analogous to how removing an `int` field means setting it to `0`. The only weird thing (as you alluded in the roundtrip tests) is that there's no way to unmarshal a slice update as `SliceSet[T](nil)` - because it will be unmarshalled as `SliceRemove[T]()` instead. But I can't think of any current uses in our codebase where the distinction would matter. Thoughts?

**Comment by jonathansharman** [comment]
@dzbee 
> It seems like we don't plan to update the repo name, is that correct? I can imagine possible reasons to not do it, but I'm curious what considerations you've given to it.

Renaming the entire repo to `github.com/nicheinc/nup` was option 5 in the PR description, but I don't think that's a viable option because it would break any existing dependent code, including non-Niche dependent code (which is a remote but real possibility since this is open-source now). I think option 1 strikes a decent balance between ergonomics and not breaking anything.

File: nup/sliceUpdate_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (dzbee) [reply nup/sliceUpdate_test.go]: I think this is okay -- I like that it's documented in the tests now. I'm still a little perturbed that there is some weirdness like `SliceSet[T](nil).String()` is `"[]"` but marshals (IMO correctly) to `null`. Ultimately what I'm concerned about is a gotcha cropping up from someone using `SliceSet[T](nil)`, for instances in logging when a variable of type `SliceUpdate` might be used as a stringer and the logs not faithfully representing data that was sent over the wire, or from some future functionality that doesn't exist yet where this could become a corner case. I can accept it with the documentation and tests as they are and the fact that `SliceSet[T](nil)` and `SliceRemove[T]()` are almost identical for all functionality and most crucially for marshalling, but it still gives me some unease.

If we're really reticent to change the `SliceSet` signature to include an error to exclude `SliceSet[T](nil)` another possibility (albeit a really aggressive one) would be to have a call w/ `nil` create a panic.

**Review by dzbee — APPROVED** [review]
LGTM

I've left one more suggestion for thinking about how to shut down the ambiguity of removing a slice to ensure a One True Way, but it's a pretty opinionated idea so I totally defer to your discretion.

File: nup/sliceUpdate_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonathansharman) [reply nup/sliceUpdate_test.go]: Essentially, we need to decide whether we want to support the `nil` value for set operations and, if not, the mechanism for disallowing it. For the sake of ergonomics, I'd really like to avoid an error return if possible. And panicking seems a little dangerous. I can think of at least two other options for disallowing `nil` slice-set operations:

1. Return the equivalent of `SliceRemove[T]()` in that case. (This would obviate the need for `SliceRemoveOrSet`.)
2. Allocate an empty array in that case, i.e. `SliceSet[T](nil)` is equivalent to `SliceSet([]T{})`.

This is a tough call. I think I'll make a Nitty Gritty topic on this, unless any of these options jumps out to you as obviously the best.

[the remark this answers is not in the document]
Comment (dzbee) [reply nup/sliceUpdate_test.go]: I like option 1. It's less magic than turning a nil slice into an empty one and I think it also would almost always reflect the user's expectation.
